lib/Conversion/CalyxToFSM/MaterializeFSM.cpp

```diff
@@ -0,0 +1,199 @@
+  };
+
+  /// Returns an operand vector for the states which should be enabled for a
+  /// given state, as well as the state done signals that should be considered
+  /// for the state transition guard. We do this here to avoid
+  /// assignStateOutputOperands/assignStateTransitionGuard having to each walk
+  /// the set of output operations.
+  StateEnableInfo gatherStateEnableInfo(StateOp stateOp) {
+    StateEnableInfo info;
+    size_t portIndex = 0;
+    auto &enabledGroups = stateEnables[stateOp];
+    for (auto group : referencedGroups) {
+      if (enabledGroups.contains(group)) {
+        info.outputOperands.push_back(c1);
+        info.doneGuards.push_back(machineOp.getArgument(portIndex));
+      } else
+        info.outputOperands.push_back(c0);
+
+      ++portIndex;
+    }
+    return info;
+  }
+
+  /// Assigns the 'fsm.output' operation of the provided 'state' to enabled the
+  /// set of provided groups. If 'topLevelDone' is set, also asserts the
+  /// top-level done signal.
+  void assignStateOutputOperands(StateOp stateOp,
+                                 llvm::SmallVector<Value> outputOperands,
+                                 bool topLevelDone = false) {
+    assert(outputOperands.size() == machineOp.getNumArguments() - 1 &&
+           "Expected exactly one value for each uniquely referenced group in "
+           "this machine");
+    // outputOperands is expected to only have
+    outputOperands.push_back(topLevelDone ? c1 : c0);
+    auto outputOp = stateOp.output().getOps<fsm::OutputOp>();
+    assert(!outputOp.empty() &&
+           "Expected an fsm.output op inside the state output region");
+    (*outputOp.begin()).getOperation()->setOperands(outputOperands);
+  }
+
+  /// Extends every `fsm.return` guard in the transitions of this state to also
+  /// include the provided set of 'doneGuards'.
+  void assignStateTransitionGuard(OpBuilder &b, StateOp stateOp,
+                                  llvm::ArrayRef<Value> doneGuards) {
+    for (auto transition : stateOp.transitions().getOps<fsm::TransitionOp>()) {
+      auto guardOp = transition.getGuardReturn();
+      llvm::SmallVector<Value> guards;
+      llvm::append_range(guards, doneGuards);
+      if (guardOp.operand())
+        guards.push_back(guardOp.operand());
+
+      if (guards.empty())
+        continue;
+
+      b.setInsertionPoint(guardOp);
+      Value guardConjunction;
+      if (guards.size() == 1)
+        guardConjunction = guards.front();
+      else
+        guardConjunction = b.create<comb::AndOp>(transition.getLoc(), guards);
+      guardOp.setOperand(guardConjunction);
+    }
+  }
+
+  /// Maintain a set of all groups referenced within this fsm.machine.
+  GroupSet referencedGroups;
+
+  /// Maintain a relation between states and the groups which they enable.
+  DenseMap<fsm::StateOp, GroupSet> stateEnables;
+
+  /// A handle to the machine under transformation.
+  MachineOp machineOp;
+
+  /// Commonly used constants.
+  Value c1, c0;
```

Comment:
Is this really cleaner than just having helper functions to create the constants, and letting another GVN pass do the cleanup? IMO, it just seems like extra things to reason about / a place where use-before-initialization may get ugly. Up to you.

Reply:
This was an attempt to avoid having a`getOrCreateConstant`-like function (found in many other conversion functions) to avoid bloating the IR with unnecessary constant ops. I think it still makes sense to define these once and then reuse, but  use-before-initialization ugliness is a valid concern. I'll take another stab at it.